src/realm/object-store/c_api/util.hpp

```diff
@@ -30,8 +32,72 @@ static inline const ObjectSchema& schema_for_table(const std::shared_ptr<Realm>&
         }
     }
 
-    // FIXME: Proper exception type.
-    throw std::logic_error{"Class not in schema"};
+    throw NoSuchTable{};
+}
+
+inline void report_type_mismatch(const SharedRealm& realm, const Table& table, ColKey col_key)
+{
+    auto& schema = schema_for_table(realm, table.get_key());
+    throw PropertyTypeMismatch{schema.name, table.get_column_name(col_key)};
+}
+
+inline void check_value_assignable(const SharedRealm& realm, const Table& table, ColKey col_key, Mixed val)
+{
+    if (val.is_null()) {
+        if (col_key.is_nullable()) {
+            return;
+        }
+        auto& schema = schema_for_table(realm, table.get_key());
+        throw NotNullableException{schema.name, table.get_column_name(col_key)};
+    }
+
+    if (val.get_type() == type_TypedLink &&
+        (col_key.get_type() == col_type_Link || col_key.get_type() == col_type_LinkList)) {
+        auto obj_link = val.get<ObjLink>();
+        if (table.get_link_target(col_key)->get_key() != obj_link.get_table_key()) {
+            report_type_mismatch(realm, table, col_key);
+        }
+    }
+    else {
+        if (ColumnType(val.get_type()) != col_key.get_type()) {
+            report_type_mismatch(realm, table, col_key);
+        }
+    }
+}
+
+inline void check_value_assignable(const List& list, Mixed val)
+{
+    auto realm = list.get_realm();
+    auto table_key = list.get_parent_table_key();
+    auto table = realm->read_group().get_table(table_key);
+    auto col_key = list.get_parent_column_key();
+    return check_value_assignable(realm, *table, col_key, val);
 }
 
+inline Mixed typed_link_to_objkey(Mixed val)
+{
+    if (!val.is_null() && val.get_type() == type_TypedLink) {
```

Comment:
I don't understand this (and it's call sites). Why are we are throwing out the Table part of this link?

Reply:
Might need a comment. The C API only deals with `realm_link_t` (which is equivalent to `ObjLink`), but link APIs in Core use plain ObjKeys (the table being implicit from context). This is just a utility to convert `ObjLink` to `ObjKey` within a `Mixed`, after having type-checked the target table.

Reply:
Added a comment